## About this Repository  
These are Jupyternotebooks for scRNAseq analysis. Comments and explainations are in the notenbooks.  
>* this [notebook](scRNAseq/MF0245_CD45_sig_pos_NEW.ipynb) takes output from [10xcell ranger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/overview) performs an initial exploratory analysis of the reads incl. doublet score calculation and then runs the Seurat SCT pipeline. The notebook als contains scirpt cells for differenctial gene expresssion calculation for each cell population (cluster) against each remainin cluster in the analysis. Resulting lists can be exported or can be further analysed with enrichR to gather information about possible cell type identities. Finally, cluster specfic markers are TIERed so you can see the top markers associtated to one cluster.
>* this[notebook](scRNAseq/harmony_integration_mt30.ipynb) merges and integrates several 10x scRNAseq datasests together and then performs the same differential gene expression analysis as outlined above.  
